bug: unable to merge branch with large data change

### Component

API Server / GraphQL

### Infrahub version

0.16.0

### Current Behavior

When you want to merge a branch with a large data change (40 devices, 4000 interfaces) you get an error message:
```
{
  "name": "ApolloError",
  "graphQLErrors": [
    {
      "message": "Unable to connect to the database",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "BranchMerge"
      ]
    }
  ],
  "protocolErrors": [],
  "clientErrors": [],
  "networkError": null,
  "message": "Unable to connect to the database"
}
```
The message only seems to occur after some time and it seems that the database server has run out of memory.

```
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "req-rsp-timeout-task"
Exception in thread "HTTP-Dispatcher" java.lang.OutOfMemoryError: Java heap space
Exception in thread "neo4j.ThroughputMonitor-1" java.lang.OutOfMemoryError: Java heap space
Uncaught error from thread [cc-discovery-actor-system-akka.io.pinned-dispatcher-8]: Java heap space, shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for ActorSystem[cc-discovery-actor-system]
java.lang.OutOfMemoryError: Java heap space
Uncaught error from thread [cc-discovery-actor-system-scheduler-1]: Java heap space, shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for ActorSystem[cc-discovery-actor-system]
java.lang.OutOfMemoryError: Java heap space
Uncaught error from thread [cc-discovery-actor-system-akka.actor.internal-dispatcher-33]: Java heap space, shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for ActorSystem[cc-discovery-actor-system]
java.lang.OutOfMemoryError: Java heap space
2024-09-24 15:46:47.127+0000 INFO  Neo4j Server shutdown initiated by request
2024-09-24 15:46:47.138+0000 INFO  Stopping...
ERROR StatusConsoleListener An exception occurred processing Appender rotatingWriter.neo4j.database.neo4j.db.query.execution.pipelined.failure.csv
 org.apache.logging.log4j.core.appender.AppenderLoggingException: java.lang.OutOfMemoryError: Java heap space
        at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:165)
        at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:134)
        at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:125)
        at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:89)
        at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:683)
        at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:641)
        at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:624)
        at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:560)
        at org.apache.logging.log4j.core.config.DefaultReliabilityStrategy.log(DefaultReliabilityStrategy.java:63)
        at org.apache.logging.log4j.core.Logger.log(Logger.java:163)
        at org.apache.logging.log4j.spi.AbstractLogger.tryLogMessage(AbstractLogger.java:2168)
        at org.apache.logging.log4j.spi.AbstractLogger.logMessageTrackRecursion(AbstractLogger.java:2122)
        at org.apache.logging.log4j.spi.AbstractLogger.logMessageSafely(AbstractLogger.java:2105)
        at org.apache.logging.log4j.spi.AbstractLogger.printf(AbstractLogger.java:2095)
        at org.neo4j.logging.log4j.RotatingLogFileWriter.printf(RotatingLogFileWriter.java:71)
        at com.neo4j.metrics.output.RotatableCsvReporter.report(RotatableCsvReporter.java:234)
        at com.neo4j.metrics.output.RotatableCsvReporter.reportMeter(RotatableCsvReporter.java:180)
        at com.neo4j.metrics.output.RotatableCsvReporter.report(RotatableCsvReporter.java:144)
        at com.codahale.metrics.ScheduledReporter.report(ScheduledReporter.java:277)
        at com.codahale.metrics.ScheduledReporter.lambda$start$0(ScheduledReporter.java:206)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.OutOfMemoryError: Java heap space
```
After this issue, you seem to be unable to restart the database and therefor the application server.

This happens using the recommended hardware requirements (in this case 12CPU, 16GB RAM).

### Expected Behavior

The merge operation should succeed!

### Steps to Reproduce

- load an instance of Infrahub with the demo schema `inv dev.start demo.load-infra-schema`
- create a branch `infrahubctl branch create test`
- run the following script to load the dataset in the branch test (40 devices, 100 interfaces each) `infrahubctl run <script.py> num_devices=40 --branch test`
```python
import logging
from infrahub_sdk import InfrahubClient


async def run(client: InfrahubClient, log: logging.Logger, branch: str, num_devices: int=50) -> None:
    site = await client.create("LocationSite", name="atl1")
    await site.save(allow_upsert=True)

    num_devices = int(num_devices)

    device_batch = await client.create_batch()
    interface_batch = await client.create_batch()

    for i in range(num_devices):
        device = await client.create("InfraDevice", name=f"atl1-test{i}", site=site, type="testing")
        device_batch.add(task=device.save, node=device, allow_upsert=True)
        log.info(f"Added device {device.name.value}")
        
    async for node, result in device_batch.execute():
        print(f"device {node.name.value} was created in Infrahub succesfully")
        client.store.set(key=node.name.value, node=node)

    for i in range(num_devices):
        for j in range(100):
            interface = await client.create("InfraInterfaceL2", name=f"Ethernet{j}", l2_mode="Access", speed=10000, device=client.store.get(key=f"atl1-test{i}"))
            interface_batch.add(task=interface.save, node=interface, allow_upsert=True)
            log.info(f"  Added interface {interface.name.value} for device {interface.device.peer.name.value}")

    async for node, result in interface_batch.execute():
        print(f"interface {node.name.value} {node.device.peer.name.value} was created in Infrahub succesfully")
```
- go to the branch detail page for the `test` branch
- merge the branch

### Additional Information

_No response_

--- a/backend/infrahub/core/branch/tasks.py
+++ b/backend/infrahub/core/branch/tasks.py
@@ -17,7 +17,7 @@ from infrahub.core.validators.models.validate_migration import SchemaValidateMig
 from infrahub.core.validators.tasks import schema_validate_migrations
 from infrahub.dependencies.registry import get_component_registry
 from infrahub.events.branch_action import BranchDeleteEvent
-from infrahub.exceptions import ValidationError
+from infrahub.exceptions import MergeFailedError, ValidationError
 from infrahub.log import get_log_data
 from infrahub.message_bus import Meta, messages
 from infrahub.services import services
@@ -138,7 +138,7 @@ async def rebase_branch(branch: str) -> None:
 
 
 @flow(name="branch-merge")
-async def merge_branch(branch: str, conflict_resolution: dict[str, bool] | None = None) -> None:
+async def merge_branch(branch: str) -> None:
     service = services.service
     log = get_run_logger()
 
@@ -150,14 +150,21 @@ async def merge_branch(branch: str, conflict_resolution: dict[str, bool] | None 
 
     merger: BranchMerger | None = None
     async with lock.registry.global_graph_lock():
-        async with service.database.start_transaction() as db:
-            diff_coordinator = await component_registry.get_component(DiffCoordinator, db=db, branch=obj)
-            diff_merger = await component_registry.get_component(DiffMerger, db=db, branch=obj)
-            merger = BranchMerger(
-                db=db, diff_coordinator=diff_coordinator, diff_merger=diff_merger, source_branch=obj, service=service
-            )
-            await merger.merge(conflict_resolution=conflict_resolution)
-            await merger.update_schema()
+        diff_coordinator = await component_registry.get_component(DiffCoordinator, db=service.database, branch=obj)
+        diff_merger = await component_registry.get_component(DiffMerger, db=service.database, branch=obj)
+        merger = BranchMerger(
+            db=service.database,
+            diff_coordinator=diff_coordinator,
+            diff_merger=diff_merger,
+            source_branch=obj,
+            service=service,
+        )
+        try:
+            await merger.merge()
+        except Exception as exc:
+            await merger.rollback()
+            raise MergeFailedError(branch_name=branch) from exc
+        await merger.update_schema()
 
     if merger and merger.migrations:
         errors = await schema_apply_migrations(

--- a/backend/infrahub/core/diff/merger/merger.py
+++ b/backend/infrahub/core/diff/merger/merger.py
@@ -4,7 +4,7 @@ from typing import TYPE_CHECKING
 
 from infrahub.core import registry
 from infrahub.core.diff.model.path import BranchTrackingId
-from infrahub.core.diff.query.merge import DiffMergePropertiesQuery, DiffMergeQuery
+from infrahub.core.diff.query.merge import DiffMergePropertiesQuery, DiffMergeQuery, DiffMergeRollbackQuery
 
 if TYPE_CHECKING:
     from infrahub.core.branch import Branch
@@ -65,3 +65,9 @@ class DiffMerger:
         self.source_branch.branched_from = at.to_string()
         await self.source_branch.save(db=self.db)
         registry.branch[self.source_branch.name] = self.source_branch
+
+    async def rollback(self, at: Timestamp) -> None:
+        rollback_query = await DiffMergeRollbackQuery.init(
+            db=self.db, branch=self.source_branch, target_branch=self.destination_branch, at=at
+        )
+        await rollback_query.execute(db=self.db)

--- a/backend/infrahub/core/diff/query/merge.py
+++ b/backend/infrahub/core/diff/query/merge.py
@@ -166,12 +166,17 @@ CALL {
             CALL {
                 WITH n, rel_name, rel_peer_id, related_rel_status
                 MATCH (n)
-                    -[source_r_rel_1:IS_RELATED {branch: $source_branch}]
+                    -[source_r_rel_1:IS_RELATED]
                     -(r:Relationship {name: rel_name})
-                    -[source_r_rel_2:IS_RELATED {branch: $source_branch}]
+                    -[source_r_rel_2:IS_RELATED]
                     -(:Node {uuid: rel_peer_id})
-                WHERE source_r_rel_1.from <= $at AND source_r_rel_1.to IS NULL
+                WHERE source_r_rel_1.branch IN [$source_branch, $target_branch]
+                AND source_r_rel_2.branch IN [$source_branch, $target_branch]
+                AND source_r_rel_1.from <= $at AND source_r_rel_1.to IS NULL
                 AND source_r_rel_2.from <= $at AND source_r_rel_2.to IS NULL
+                WITH n, rel_name, rel_peer_id, related_rel_status, r, source_r_rel_1, source_r_rel_2
+                ORDER BY source_r_rel_1.branch_level DESC, source_r_rel_2.branch_level DESC, source_r_rel_1.from DESC, source_r_rel_2.from DESC
+                LIMIT 1
                 RETURN r, CASE
                     WHEN startNode(source_r_rel_1).uuid = n.uuid THEN "r"
                     ELSE "l"
@@ -431,5 +436,46 @@ CALL {
         }
     }
 }
+        """
+        self.add_to_query(query=query)
+
+
+class DiffMergeRollbackQuery(Query):
+    name = "diff_merge_rollback"
+    type = QueryType.WRITE
+    insert_return = False
+
+    def __init__(
+        self,
+        at: Timestamp,
+        target_branch: Branch,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.at = at
+        self.target_branch = target_branch
+        self.source_branch_name = self.branch.name
+
+    async def query_init(self, db: InfrahubDatabase, **kwargs: Any) -> None:
+        self.params = {
+            "at": self.at.to_string(),
+            "target_branch": self.target_branch.name,
+            "source_branch": self.source_branch_name,
+        }
+        query = """
+        // ---------------------------
+        // reset to times on target branch
+        // ---------------------------
+        CALL {
+            OPTIONAL MATCH ()-[r_to {to: $at, branch: $target_branch}]-()
+            SET r_to.to = NULL
+        }
+        // ---------------------------
+        // reset from times on target branch
+        // ---------------------------
+        CALL {
+            OPTIONAL MATCH ()-[r_from {from: $at, branch: $target_branch}]-()
+            DELETE r_from
+        }
         """
         self.add_to_query(query=query)

--- a/backend/infrahub/core/merge.py
+++ b/backend/infrahub/core/merge.py
@@ -44,6 +44,7 @@ class BranchMerger:
         self.diff_coordinator = diff_coordinator
         self.diff_merger = diff_merger
         self.migrations: list[SchemaUpdateMigrationInfo] = []
+        self._merge_at = Timestamp()
         self._graph_diff: Optional[BranchDiffer] = None
 
         self._source_schema: Optional[SchemaBranch] = None
@@ -230,7 +231,6 @@ class BranchMerger:
     async def merge(
         self,
         at: Optional[Union[str, Timestamp]] = None,
-        conflict_resolution: Optional[dict[str, bool]] = None,  # pylint: disable=unused-argument
     ) -> None:
         """Merge the current branch into main."""
         if self.source_branch.name == registry.default_branch:
@@ -252,9 +252,12 @@ class BranchMerger:
 
         # TODO need to find a way to properly communicate back to the user any issue that could come up during the merge
         # From the Graph or From the repositories
-        at = Timestamp(at)
-        await self.diff_merger.merge_graph(at=at)
+        self._merge_at = Timestamp(at)
+        await self.diff_merger.merge_graph(at=self._merge_at)
         await self.merge_repositories()
+
+    async def rollback(self) -> None:
+        await self.diff_merger.rollback(at=self._merge_at)
 
     async def merge_repositories(self) -> None:
         # Collect all Repositories in Main because we'll need the commit in Main for each one.

--- a/backend/infrahub/exceptions.py
+++ b/backend/infrahub/exceptions.py
@@ -317,3 +317,11 @@ class HTTPServerTimeoutError(HTTPServerError):
 
 class HTTPServerSSLError(HTTPServerError):
     HTTP_CODE = 503
+
+
+class MergeFailedError(Error):
+    HTTP_CODE: int = 500
+
+    def __init__(self, branch_name: str) -> None:
+        self.message = f"Failed to merge branch '{branch_name}'"
+        super().__init__(self.message)

--- a/backend/infrahub/graphql/mutations/proposed_change.py
+++ b/backend/infrahub/graphql/mutations/proposed_change.py
@@ -140,7 +140,6 @@ class InfrahubProposedChangeMutation(InfrahubMutationMixin, Mutation):
             )
 
             if updated_state == ProposedChangeState.MERGED:
-                conflict_resolution: dict[str, bool] = {}
                 source_branch = await Branch.get_by_name(db=dbt, name=proposed_change.source_branch.value)
                 validations = await proposed_change.validations.get_peers(db=dbt)
                 for validation in validations.values():
@@ -158,13 +157,10 @@ class InfrahubProposedChangeMutation(InfrahubMutationMixin, Mutation):
                                 raise ValidationError(
                                     "Data conflicts found on branch and missing decisions about what branch to keep"
                                 )
-                            if check.conflicts.value:
-                                keep_source_value = check.keep_branch.value.value == "source"
-                                conflict_resolution[check.conflicts.value[0]["path"]] = keep_source_value
 
                 await context.service.workflow.execute_workflow(
                     workflow=BRANCH_MERGE,
-                    parameters={"branch": source_branch.name, "conflict_resolution": conflict_resolution},
+                    parameters={"branch": source_branch.name},
                 )
 
         return proposed_change, result

--- a/backend/tests/integration/diff/test_merge_rollback.py
+++ b/backend/tests/integration/diff/test_merge_rollback.py
@@ -1,0 +1,157 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+from unittest.mock import patch
+
+import pytest
+from infrahub_sdk.exceptions import GraphQLError
+
+from infrahub.core.initialization import create_branch
+from infrahub.core.manager import NodeManager
+from infrahub.core.merge import BranchMerger
+from infrahub.core.node import Node
+from infrahub.services.adapters.cache.redis import RedisCache
+from tests.constants import TestKind
+from tests.helpers.schema import CAR_SCHEMA, load_schema
+from tests.helpers.test_app import TestInfrahubApp
+
+if TYPE_CHECKING:
+    from infrahub_sdk import InfrahubClient
+
+    from infrahub.core.branch.models import Branch
+    from infrahub.database import InfrahubDatabase
+    from tests.adapters.message_bus import BusSimulator
+
+
+BRANCH_MERGE = """
+mutation($branch: String!) {
+    BranchMerge(data: { name: $branch }) {
+        ok
+    }
+}
+"""
+
+
+class BrokenBranchMerger:
+    def __init__(self, *args, **kwargs) -> None:
+        self.real_merger = BranchMerger(*args, **kwargs)
+
+    async def merge(self, at=None) -> None:
+        await self.real_merger.merge(at=at)
+        raise ValueError("This is broken on purpose")
+
+    async def rollback(self) -> None:
+        await self.real_merger.rollback()
+
+
+class TestBranchMergeRollback(TestInfrahubApp):
+    @pytest.fixture(scope="class")
+    async def initial_dataset(
+        self,
+        db: InfrahubDatabase,
+        initialize_registry: None,
+        bus_simulator: BusSimulator,
+        prefect_test_fixture: None,
+    ) -> dict[str, Node]:
+        await load_schema(db, schema=CAR_SCHEMA)
+
+        bus_simulator.service.cache = RedisCache()
+
+        john = await Node.init(schema=TestKind.PERSON, db=db)
+        await john.new(db=db, name="John", height=175, description="The famous Joe Doe")
+        await john.save(db=db)
+        kara = await Node.init(schema=TestKind.PERSON, db=db)
+        await kara.new(db=db, name="Kara Thrace", height=165, description="Starbuck")
+        await kara.save(db=db)
+        murphy = await Node.init(schema=TestKind.PERSON, db=db)
+        await murphy.new(db=db, name="Alex Murphy", height=185, description="Robocop")
+        await murphy.save(db=db)
+        omnicorp = await Node.init(schema=TestKind.MANUFACTURER, db=db)
+        await omnicorp.new(db=db, name="Omnicorp", customers=[murphy])
+        await omnicorp.save(db=db)
+        cyberdyne = await Node.init(schema=TestKind.MANUFACTURER, db=db)
+        await cyberdyne.new(db=db, name="Cyberdyne")
+        await cyberdyne.save(db=db)
+
+        t_800 = await Node.init(schema=TestKind.CAR, db=db)
+        await t_800.new(
+            db=db,
+            name="Cyberdyne systems model 101",
+            color="Chrome",
+            description="killing machine with secret heart of gold",
+            owner=john,
+            manufacturer=cyberdyne,
+        )
+        await t_800.save(db=db)
+        ed_209 = await Node.init(schema=TestKind.CAR, db=db)
+        await ed_209.new(
+            db=db,
+            name="ED-209",
+            color="Chrome",
+            description="still working on doing stairs",
+            owner=murphy,
+            manufacturer=omnicorp,
+        )
+        await ed_209.save(db=db)
+
+        return {
+            "john": john,
+            "kara": kara,
+            "murphy": murphy,
+            "omnicorp": omnicorp,
+            "cyberdyne": cyberdyne,
+            "t_800": t_800,
+            "ed_209": ed_209,
+        }
+
+    @pytest.fixture(scope="class")
+    async def branch1(self, db: InfrahubDatabase) -> Branch:
+        return await create_branch(db=db, branch_name="branch1")
+
+    @pytest.fixture(scope="class")
+    async def branch1_data(self, db: InfrahubDatabase, initial_dataset: dict[str, Node], branch1: Branch) -> dict[str, Node]:
+        kara_branch = await NodeManager.get_one(db=db, branch=branch1, id=initial_dataset["kara"].id)
+        await kara_branch.delete(db=db)
+
+        sarah = await Node.init(schema=TestKind.PERSON, db=db, branch=branch1)
+        await sarah.new(db=db, name="Sarah", height=161, description="no fate")
+        await sarah.save(db=db)
+
+        t_800_branch = await NodeManager.get_one(db=db, branch=branch1, id=initial_dataset["t_800"].id)
+        await t_800_branch.owner.update(db=db, data=sarah)
+        await t_800_branch.save(db=db)
+
+        ocp_branch = await NodeManager.get_one(db=db, branch=branch1, id=initial_dataset["omnicorp"].id)
+        ocp_branch.name.value = "Omni Consumer Products"
+        await ocp_branch.save(db=db)
+
+        return {"sarah": sarah}
+
+    async def test_merge_branch_rollback(
+        self,
+        db: InfrahubDatabase,
+        client: InfrahubClient,
+        initial_dataset: dict[str, Node],
+        branch1: Branch,
+        branch1_data: dict[str, Node],
+    ) -> None:
+        with patch("infrahub.core.branch.tasks.BranchMerger", new=BrokenBranchMerger):
+            with pytest.raises(GraphQLError) as exc:
+                await client.execute_graphql(query=BRANCH_MERGE, variables={"branch": branch1.name})
+
+            assert exc
+            assert f"Failed to merge branch '{branch1.name}'" in exc.value.message
+
+        # check that the changes on the branch have all been rolled back
+        kara_main = await NodeManager.get_one(db=db, id=initial_dataset["kara"].id)
+        assert kara_main.id
+
+        sarah = await NodeManager.get_one(db=db, id=branch1_data["sarah"].id)
+        assert sarah is None
+
+        t_800_main = await NodeManager.get_one(db=db, id=initial_dataset["t_800"].id)
+        owner_peer = await t_800_main.owner.get_peer(db=db)
+        assert owner_peer.id == initial_dataset["john"].id
+
+        ocp_main = await NodeManager.get_one(db=db, id=initial_dataset["omnicorp"].id)
+        assert ocp_main.name.value == "Omnicorp"

--- a/backend/tests/integration/diff/test_merge_rollback.py
+++ b/backend/tests/integration/diff/test_merge_rollback.py
@@ -109,7 +109,9 @@ class TestBranchMergeRollback(TestInfrahubApp):
         return await create_branch(db=db, branch_name="branch1")
 
     @pytest.fixture(scope="class")
-    async def branch1_data(self, db: InfrahubDatabase, initial_dataset: dict[str, Node], branch1: Branch) -> dict[str, Node]:
+    async def branch1_data(
+        self, db: InfrahubDatabase, initial_dataset: dict[str, Node], branch1: Branch
+    ) -> dict[str, Node]:
         kara_branch = await NodeManager.get_one(db=db, branch=branch1, id=initial_dataset["kara"].id)
         await kara_branch.delete(db=db)
 

--- a/backend/tests/unit/core/diff/test_diff_and_merge.py
+++ b/backend/tests/unit/core/diff/test_diff_and_merge.py
@@ -74,10 +74,11 @@ class TestDiffAndMerge:
             db=db, branch=branch2, schema=schema_branch, limit=["TestCar", "TestPerson"], update_db=True
         )
 
+        at = Timestamp()
         diff_coordinator = await self._get_diff_coordinator(db=db, branch=branch2)
         await diff_coordinator.update_branch_diff(base_branch=default_branch, diff_branch=branch2)
         diff_merger = await self._get_diff_merger(db=db, branch=branch2)
-        await diff_merger.merge_graph(at=Timestamp())
+        await diff_merger.merge_graph(at=at)
 
         updated_schema = await registry.schema.load_schema_from_db(db=db, branch=default_branch)
         car_schema_main = updated_schema.get(name="TestCar", duplicate=False)
@@ -87,6 +88,15 @@ class TestDiffAndMerge:
         assert new_bool_attr.default_value is False
         new_str_attr = car_schema_main.get_attribute(name="nickname")
         assert new_str_attr.default_value == "car"
+
+        await diff_merger.rollback(at=at)
+
+        rolled_back_schema = await registry.schema.load_schema_from_db(db=db, branch=default_branch)
+        car_schema_main = rolled_back_schema.get(name="TestCar", duplicate=False)
+        attribute_names = car_schema_main.attribute_names
+        assert "num_cupholders" not in attribute_names
+        assert "is_cool" not in attribute_names
+        assert "nickname" not in attribute_names
 
     @pytest.mark.parametrize(
         "conflict_selection,expected_value",
@@ -112,6 +122,7 @@ class TestDiffAndMerge:
         john_branch.name.value = "John-branch"
         await john_branch.save(db=db)
 
+        at = Timestamp()
         diff_coordinator = await self._get_diff_coordinator(db=db, branch=branch2)
         enriched_diff = await diff_coordinator.update_branch_diff(base_branch=default_branch, diff_branch=branch2)
         conflicts_map = enriched_diff.get_all_conflicts()
@@ -119,10 +130,15 @@ class TestDiffAndMerge:
         conflict = next(iter(conflicts_map.values()))
         await diff_repository.update_conflict_by_id(conflict_id=conflict.uuid, selection=conflict_selection)
         diff_merger = await self._get_diff_merger(db=db, branch=branch2)
-        await diff_merger.merge_graph(at=Timestamp())
+        await diff_merger.merge_graph(at=at)
 
         updated_john = await NodeManager.get_one(db=db, id=person_john_main.id)
         assert updated_john.name.value == expected_value
+
+        await diff_merger.rollback(at=at)
+
+        rolled_back_john = await NodeManager.get_one(db=db, id=person_john_main.id)
+        assert rolled_back_john.name.value == "John-main"
 
     @pytest.mark.parametrize(
         "conflict_selection",
@@ -148,6 +164,7 @@ class TestDiffAndMerge:
         await car_branch.owner.update(db=db, data=person_jane_main)
         await car_branch.save(db=db)
 
+        at = Timestamp()
         diff_coordinator = await self._get_diff_coordinator(db=db, branch=branch2)
         enriched_diff = await diff_coordinator.update_branch_diff(base_branch=default_branch, diff_branch=branch2)
         conflicts_map = enriched_diff.get_all_conflicts()
@@ -155,7 +172,7 @@ class TestDiffAndMerge:
         conflict = next(iter(conflicts_map.values()))
         await diff_repository.update_conflict_by_id(conflict_id=conflict.uuid, selection=conflict_selection)
         diff_merger = await self._get_diff_merger(db=db, branch=branch2)
-        await diff_merger.merge_graph(at=Timestamp())
+        await diff_merger.merge_graph(at=at)
 
         updated_car = await NodeManager.get_one(db=db, id=car_accord_main.id)
         owner_rel = await updated_car.owner.get(db=db)
@@ -163,6 +180,12 @@ class TestDiffAndMerge:
             assert owner_rel.peer_id == person_alfred_main.id
         if conflict_selection is ConflictSelection.DIFF_BRANCH:
             assert owner_rel.peer_id == person_jane_main.id
+
+        await diff_merger.rollback(at=at)
+
+        rolled_back_car = await NodeManager.get_one(db=db, id=car_accord_main.id)
+        owner_rel = await rolled_back_car.owner.get(db=db)
+        assert owner_rel.peer_id == person_alfred_main.id
 
     @pytest.mark.parametrize(
         "conflict_selection",
@@ -187,6 +210,7 @@ class TestDiffAndMerge:
         john_branch.name.source = person_jane_main
         await john_branch.save(db=db)
 
+        at = Timestamp()
         diff_coordinator = await self._get_diff_coordinator(db=db, branch=branch2)
         enriched_diff = await diff_coordinator.update_branch_diff(base_branch=default_branch, diff_branch=branch2)
         conflicts_map = enriched_diff.get_all_conflicts()
@@ -194,7 +218,7 @@ class TestDiffAndMerge:
         conflict = next(iter(conflicts_map.values()))
         await diff_repository.update_conflict_by_id(conflict_id=conflict.uuid, selection=conflict_selection)
         diff_merger = await self._get_diff_merger(db=db, branch=branch2)
-        await diff_merger.merge_graph(at=Timestamp())
+        await diff_merger.merge_graph(at=at)
 
         updated_john = await NodeManager.get_one(db=db, id=person_john_main.id, include_source=True)
 
@@ -203,6 +227,12 @@ class TestDiffAndMerge:
             assert attr_source.id == person_alfred_main.id
         if conflict_selection is ConflictSelection.DIFF_BRANCH:
             assert attr_source.id == person_jane_main.id
+
+        await diff_merger.rollback(at=at)
+
+        rolled_back_john = await NodeManager.get_one(db=db, id=person_john_main.id, include_source=True)
+        attr_source = await rolled_back_john.name.get_source(db=db)
+        assert attr_source.id == person_alfred_main.id
 
     @pytest.mark.parametrize(
         "conflict_selection",
@@ -228,6 +258,7 @@ class TestDiffAndMerge:
         await car_branch.owner.update(db=db, data={"id": person_john_main.id, "_relation__owner": person_jane_main.id})
         await car_branch.save(db=db)
 
+        at = Timestamp()
         diff_coordinator = await self._get_diff_coordinator(db=db, branch=branch2)
         enriched_diff = await diff_coordinator.update_branch_diff(base_branch=default_branch, diff_branch=branch2)
         conflicts_map = enriched_diff.get_all_conflicts()
@@ -236,7 +267,7 @@ class TestDiffAndMerge:
         for conflict in conflicts_map.values():
             await diff_repository.update_conflict_by_id(conflict_id=conflict.uuid, selection=conflict_selection)
         diff_merger = await self._get_diff_merger(db=db, branch=branch2)
-        await diff_merger.merge_graph(at=Timestamp())
+        await diff_merger.merge_graph(at=at)
 
         updated_car = await NodeManager.get_one(db=db, id=car_accord_main.id, include_owner=True)
         owner_rel = await updated_car.owner.get(db=db)
@@ -245,3 +276,10 @@ class TestDiffAndMerge:
             assert owner_prop.id == person_alfred_main.id
         if conflict_selection is ConflictSelection.DIFF_BRANCH:
             assert owner_prop.id == person_jane_main.id
+
+        await diff_merger.rollback(at=at)
+
+        rolled_back_car = await NodeManager.get_one(db=db, id=car_accord_main.id, include_owner=True)
+        owner_rel = await rolled_back_car.owner.get(db=db)
+        owner_prop = await owner_rel.get_owner(db=db)
+        assert owner_prop.id == person_alfred_main.id

--- a/backend/tests/unit/core/diff/test_diff_merger.py
+++ b/backend/tests/unit/core/diff/test_diff_merger.py
@@ -356,6 +356,13 @@ class TestMergeDiff:
         car_element = car_elements_by_peer_id[car_node_branch.id]
         assert car_element.source_id == car_node_branch.id
 
+        await diff_merger.rollback(at=at)
+
+        retrieved_person = await NodeManager.get_one(db=db, id=person_node_branch.id, branch=default_branch)
+        assert retrieved_person is None
+        retrieved_car = await NodeManager.get_one(db=db, id=car_node_branch.id, branch=default_branch)
+        assert retrieved_car is None
+
     @pytest.mark.parametrize("check_idempotent", [False, True])
     async def test_merge_node_deleted(
         self,
@@ -390,6 +397,15 @@ class TestMergeDiff:
 
         with pytest.raises(NodeNotFoundError):
             await NodeManager.get_one(db=db, branch=default_branch, id=person_node_main.id, raise_on_error=True)
+
+        await diff_merger.rollback(at=at)
+
+        rolled_back_person = await NodeManager.get_one(db=db, branch=default_branch, id=person_node_main.id)
+        assert rolled_back_person.id == person_node_main.id
+        assert rolled_back_person.height.value == person_node_main.height.value
+        rolled_back_car = await NodeManager.get_one(db=db, branch=default_branch, id=car_node_main.id)
+        owner_rel = await rolled_back_car.owner.get(db=db)
+        assert owner_rel.peer_id == person_node_main.id
 
     @pytest.mark.parametrize(
         "conflict_selection,expect_deleted",
@@ -433,6 +449,11 @@ class TestMergeDiff:
             target_person = await NodeManager.get_one(db=db, branch=default_branch, id=person_node_branch.id)
             assert target_person.id == person_node_branch.id
             assert target_person.get_updated_at() < at
+
+        await diff_merger.rollback(at=at)
+
+        rolled_back_person = await NodeManager.get_one(db=db, branch=default_branch, id=person_node_main.id)
+        assert rolled_back_person.id == person_node_branch.id
 
     @pytest.fixture
     def updated_person_node_diff(
@@ -668,3 +689,22 @@ class TestMergeDiff:
         assert rel_owner_node.id == car_node_main.id
         rel_source_node = await owner_rel.get_source(db=db)
         assert rel_source_node.id == person_node_main.id
+
+        await diff_merger.rollback(at=at)
+
+        rolled_back_person = await NodeManager.get_one(db=db, branch=default_branch, id=person_node_main.id)
+        assert rolled_back_person.name.value == "Albert"
+        assert rolled_back_person.height.value == 172
+        owner_attr = await rolled_back_person.height.get_owner(db=db)
+        assert owner_attr is None
+        rolled_back_car = await NodeManager.get_one(
+            db=db, branch=default_branch, id=car_node_main.id, include_owner=True, include_source=True
+        )
+        owner_rel = await rolled_back_car.owner.get(db=db)
+        assert owner_rel.peer_id == person_node_main.id
+        assert owner_rel.is_protected is False
+        assert owner_rel.is_visible is True
+        owner_prop = await owner_rel.get_owner(db=db)
+        assert owner_prop.id == car_node_main2.id
+        source_prop = await owner_rel.get_source(db=db)
+        assert source_prop.id == person_node_main2.id

